### PR TITLE
Enable address vectors to not be shared by endpoints

### PR DESCRIPTION
**Pull request by jhh67**
Some providers, such as `EFA`, do not allow address vectors to be shared across endpoints. In general we want to share address vectors to reduce overhead. If address vectors cannot be shared then an address vector must be created for each endpoint. It's complicated somewhat by scalable endpoints. In that case there are multiple transmit contexts for the same endpoint, but the transmit contexts can share the same address vector because there is only a single endpoint.

Resolves cray/chapel-private#1376 and closes cray/chapel-private#2460.

### CODE REVIEW

File: runtime/src/comm/ofi/comm-ofi.c [hunk 1]
```diff
@@ -2125,14 +2155,25 @@ void init_ofiEp(void) {
     avAttr.count *= numTxCtxs;
   }
 
-  OFI_CHK(fi_av_open(ofi_domain, &avAttr, &ofi_av, NULL));
+  // The EFA provider does not allow AVs to be shared among
+  // endpoints. For all other providers we'll share them to reduce
+  // overhead.
+
+  chpl_bool useSharedAv = !providerInUse(provType_efa);
+  if (useSharedAv) {
```
Comment (gbtitus) [inline runtime/src/comm/ofi/comm-ofi.c:2163]: Further down at (new) line 2275 the control flow is opposite: an outer if-stmt based on `useScalEp` surrounding an inner if-stmt based on `ofi_av == NULL`. That made this a little hard to follow for me, and I'm still not totally convinced it's right, especially if the cxi provider someday supported scalable endpoints. Maybe just take the hit on code redundancy, and put whatever control flow we need for AVs into both branches of the `if (useScalEp)` below, instead of doing it on its own here?

**Review by gbtitus — CHANGES_REQUESTED** [review]
There are some things that need fixing or at least cleanup here, but the overall changes look good!

File: runtime/src/comm/ofi/comm-ofi.c [hunk 2]
```diff
@@ -2125,14 +2155,25 @@ void init_ofiEp(void) {
     avAttr.count *= numTxCtxs;
   }
 
-  OFI_CHK(fi_av_open(ofi_domain, &avAttr, &ofi_av, NULL));
+  // The EFA provider does not allow AVs to be shared among
+  // endpoints. For all other providers we'll share them to reduce
+  // overhead.
+
+  chpl_bool useSharedAv = !providerInUse(provType_efa);
```
[the remark this answers is not in the document]
Comment (jhh67) [reply runtime/src/comm/ofi/comm-ofi.c:2162]: I added some comments to help clarify the control flow. I didn't make the change you suggested because you have to create the AV before the EP.

[the remark this answers is not in the document]
Comment (gbtitus) [reply runtime/src/comm/ofi/comm-ofi.c:2168]: Okay.  I  would have ordered the code a little differently, but I'll defer to author style.
I do think that here, as in the case further down, we should be using `DBG_AV` (newly created) instead of `DBG_TCIPS`. `TCIPS` is really for things related specifically to tx context info pointers, i.e., the `tciTab` and I see this as having more to do with address vectors.

**Review by gbtitus — APPROVED** [review]
👍
